File: Assignment4/task2.scala

import org.apache.spark.{SparkConf, SparkContext}
import collection.mutable.{HashMap, ListBuffer}
import scala.math.BigDecimal
import java.io._
import scala.collection.mutable
object task2 {

  def main(args: Array[String]): Unit={
    val sparkConf = new SparkConf().setAppName("assign-4-task2").setMaster("local[*]").set("spark.executor.memory", "4g").set("spark.driver.memory", "4g")
    val sc = new SparkContext(sparkConf)
    sc.setLogLevel("ERROR")

/*
    val threshold = 7
    val input_file_path = "/Users/gopi/Desktop/Assignment4/ub_sample_data.csv"
    val betweenFilePath = "/Users/gopi/Desktop/Assignment4/scala/between_scala.csv"
    val outputFilePath = "/Users/gopi/Desktop/Assignment4/scala/output_2_scala.csv"
*/


    println("Arguments are:")
    args.foreach(println)
    val threshold = args(0).toInt
    val input_file_path = args(1)
    val betweenFilePath = args(2)
    val outputFilePath = args(3)

    val initial_rdd = sc.textFile(input_file_path)
    val first = initial_rdd.first()
    println(first)
    val grouped = initial_rdd.filter(line => line != first).map(_.split(",")).map(line => (line(0), line(1))).groupByKey()
      .map(line => (line._1, line._2.toList.sortBy(x => x).toSet)).collectAsMap()
    val keySet = grouped.keySet
    val combinations = keySet.subsets(2).map(_.toList)
    val adjacencyList = new HashMap[String, mutable.Set[String]] with mutable.MultiMap[String, String]
    val adjacencyListCopy = new HashMap[String, mutable.Set[String]] with mutable.MultiMap[String, String]
    var dataSet = Set[String]()

    for(user_ids <- combinations) {
      val set1: Set[String] = grouped(user_ids(0))
      val set2: Set[String] = grouped(user_ids(1))

      val intersect = set1.intersect(set2)
      if (intersect.size >= threshold) {
        adjacencyList.addBinding(user_ids(0), user_ids(1))
        adjacencyList.addBinding(user_ids(1), user_ids(0))
        adjacencyListCopy.addBinding(user_ids(0), user_ids(1))
        adjacencyListCopy.addBinding(user_ids(1), user_ids(0))
        var key = getKey(user_ids(0),  user_ids(1))
        dataSet += key
      }
    }
    //adjacencyList.foreach(println)
    //println(adjacencyList.size)
    var betweeness = getBetweeness(adjacencyList)
    writeBetweenessToFile(betweeness, betweenFilePath)
    var m = betweeness.size

    var communities = getCommunities(adjacencyList)
    var modularity = getModularity(adjacencyList, communities, m, dataSet)
    var totalEdges = betweeness.size

    while(totalEdges > 0) {
      var largeEdgeValue = betweeness(0)._2
      var largeEdge = betweeness(0)._1

      for( edge <- betweeness) {
        val key = edge._1
        if(edge._2 == largeEdgeValue) {
          var v1 = getVertexFromKey(key)(0)
          var v2 = getVertexFromKey(key)(1)
          adjacencyListCopy(v1).remove(v2)
          adjacencyListCopy(v2).remove(v1)
          totalEdges -= 1
        }
      }

      var modifiedCommunities = getCommunities(adjacencyListCopy)
      var modifiedModularity = getModularity(adjacencyList, modifiedCommunities, m, dataSet)

      if(modifiedModularity >= modularity) {
        communities = modifiedCommunities
        modularity = modifiedModularity
      }
      betweeness = getBetweeness(adjacencyListCopy)
    }

    communities = communities.sortWith((a,b) => a.size < b.size || (a.size == b.size && a(0) < b(0)))
    writeCommunitiesToFile(communities, outputFilePath)
    //communities.foreach(println)
    //println(communities.size)
  }


  def  getKey(s1: String, s2:String) : String={
    var key = ""
    if(s1 < s2) {
      key = s1 + ":" + s2
    } else {
      key = s2 + ":" + s1
    }
    return key
  }

  def getVertexFromKey(key : String): Array[String] ={
    return key.split(":")

  }
  def getBetweeness(adjacencyList: HashMap[String, mutable.Set[String]]) : List[(String, Double)]={
    val vertices = adjacencyList.keySet.toList
    var results = new HashMap[String, Double].withDefaultValue(0.0)
    val counts = Array.fill(20){0}
    //print(counts)
    for(vertex <- vertices){
      val neighbours = mutable.Queue[String]()
      var nodesInQueue = Set[String]()
      neighbours.enqueue(vertex)
      nodesInQueue += vertex

      val childParentMap = new HashMap[String, mutable.Set[String]] with mutable.MultiMap[String, String]
      val shortestPathCount = new HashMap[String, Double].withDefaultValue(0.0)
      shortestPathCount(vertex) = 1.0
      val levelsMap = new HashMap[String, Int].withDefaultValue(0)
      levelsMap(vertex) = 0

      while(!neighbours.isEmpty) {
        val currentNode = neighbours.dequeue()
        //nodesInQueue -= currentNode
        val nextLevelNodes = adjacencyList(currentNode)
        for(neighbour <- nextLevelNodes){
          if(!levelsMap.contains(neighbour)) {
            levelsMap(neighbour) = levelsMap(currentNode) + 1
            shortestPathCount(neighbour) = 1.0

            if (!nodesInQueue.contains(neighbour)){
              neighbours.enqueue(neighbour)
              nodesInQueue += neighbour
            }
            childParentMap.addBinding(neighbour, currentNode)
          } else {
            if(levelsMap(neighbour) ==  levelsMap(currentNode) + 1) {
              childParentMap.addBinding(neighbour, currentNode)
              shortestPathCount(neighbour) = shortestPathCount(neighbour) + 1.0
            }
          }
        }
      }

      var set = levelsMap.values.toSet
      //println(set.size)
      counts(set.size) = counts(set.size) + 1

      //println(nodesInQueue.size)
      var nodeScore = new HashMap[String, Double].withDefaultValue(1.0)
      var betweeness = new HashMap[String, Double]

      val sortedVertices = levelsMap.toSeq.sortWith(_._2 > _._2).toList
      for(sortedVertex <- sortedVertices){
        var node = sortedVertex._1
        if(node != vertex){
          val parents = childParentMap(node)
          var share = 0.0
          for(parent <- parents){
            share = share + shortestPathCount(parent)
          }
          //println(share)
          for(parent <- parents) {
            val key = Array(parent, node).toSeq.sortBy(x => x).toList
            val key1 = key(0) + ":" + key(1)
            betweeness(key1) = (nodeScore(node) * shortestPathCount(parent)) / share
            nodeScore(parent) = nodeScore(parent) + betweeness(key1)
          }
        }
      }
      //betweeness.foreach(println)

      for(edge <- betweeness.keySet) {
        results(edge) = results(edge) + (betweeness(edge) / 2.0)
      }

    }
    val finalResults = results.toSeq.sortWith((a,b) => a._2 > b._2).toList
    //println(finalResults(0))
    //println(finalResults.size)
    //println(finalResults)
    //counts.foreach(println)
    return finalResults
  }

  def getCommunities(adjacencyList: HashMap[String, mutable.Set[String]]) : List[List[String]]={
    var communities = List[List[String]]()
    var vertices = adjacencyList.keySet
    //println(vertices.size)
    while(vertices.size > 0){
      var verticesList = vertices.toList
      var firstNode = verticesList(0)
      var visitedNodes = Set[String]()

      var neighbours = mutable.Queue[String]()
      var nodesInQueue = Set[String]()
      neighbours.enqueue(firstNode)
      nodesInQueue += firstNode

      while(!neighbours.isEmpty) {
        var vertex = neighbours.dequeue()
        visitedNodes += vertex

        var adjacentNodes = adjacencyList(vertex)
        for(adjacentNode <- adjacentNodes) {
          if (!visitedNodes.contains(adjacentNode) && !nodesInQueue.contains(adjacentNode)) {
            neighbours.enqueue(adjacentNode)
            nodesInQueue += adjacentNode
          }
        }
      }

      var community = visitedNodes.toSeq.sortWith((a,b) => a < b).toList
      communities = communities :+ community
      vertices = vertices.diff(visitedNodes)
    }
    return communities
  }

  def getModularity(adjacencyList: HashMap[String, mutable.Set[String]], communities : List[List[String]], m : Int, dataSet : Set[String]): Double = {

    var result = 0.0
    for(community <- communities) {
      for(node1 <- community){
        for(node2 <- community) {
          var key  = getKey(node1, node2)

          var length1 = adjacencyList(node1).size
          var length2 = adjacencyList(node2).size
          var value = 0
          if(dataSet.contains(key)) {
            value = 1
          }

          val communityModularity = value - ((length1 * length2) / (2.0 * m))
          //println(communityModularity)
          result += communityModularity
        }
      }
    }
    return result / (2.0 * m)
  }

  def writeCommunitiesToFile(results: List[List[String]], outputFilePath: String) : Unit={
    val pw = new PrintWriter(new File(outputFilePath))
    for(community <- results) {
      var string = ""
      for(i <- 0 until(community.size)) {
        //println("community is:",community(i))
        if(i != community.size-1) {
          string = string + "'" + community(i) + "', "
        } else {
          string = string + "'" + community(i) + "'"
        }
      }
      //println("String is :", string)
      pw.write(string)
      pw.write("\n")
    }
    pw.close()
  }

  def writeBetweenessToFile(resultsUnsorted: List[(String, Double)], outputFilePath: String) : Unit={
    //var results = resultsUnsorted.toSeq.sortWith((a,b) => a._2 > b._2 || (a._2 == b._2 && a._1(0) < b._1(0)) || (a._2 == b._2 && a._1(0) == b._1(0) && a._1(1) < b._1(1))).toList
    var results = getBetweenessStrings(resultsUnsorted)
    val pw = new PrintWriter(new File(outputFilePath))
    for(community <- results) {
      var string = ""
      var b = BigDecimal(community._3).setScale(5, BigDecimal.RoundingMode.HALF_UP).toString
      b = stripZero(b)
      string = "('" + community._1 + "', '" + community._2 + "')," + b
      pw.write(string)
      pw.write("\n")
    }
    pw.close()
  }

  def getBetweenessStrings(results : List[(String, Double)]) : List[(String, String, Double)]={
    var finalData = ListBuffer[(String, String, Double)]()
    finalData.foreach(println)
    for(result <- results) {
      var str = result._1.split(":")
      finalData += ((str(0), str(1), result._2))
    }
    //val x = finalData.sortWith((a,b) => (a._3 > b._3) || (a._3 == b._3 && a._1 < b._1)).toList
    var x = finalData.sortBy(data => (-data._3, data._1)).toList
    return x
  }

  def stripZero(b : String) : String={
    val index = b.indexOf(".")
    var finalString = ""
    var substring = b.substring(index+1, b.size)
    var size = substring.size-1
    var trailingCount = 0
    if(substring(size) == '0') {
      trailingCount+=1
    }
    if(substring(size-1) == '0' && trailingCount ==1) {
      trailingCount += 1
    }
    if(substring(size-2) == '0' && trailingCount ==2) {
      trailingCount += 1
    }
    if(substring(size-3) == '0' && trailingCount ==3) {
      trailingCount += 1
    }
    if(substring(size-4) == '0' && trailingCount ==4) {
      trailingCount += 1
    }
    if(trailingCount== 1) {
      finalString = b.substring(0, b.size-1)
    }

    if(trailingCount == 2) {
      finalString = b.substring(0, b.size-2)
    }
    if(trailingCount == 3) {
      finalString = b.substring(0, b.size-3)
    }
    if(trailingCount == 4) {
      finalString = b.substring(0, b.size-4)
    }
    if(trailingCount == 5) {
      finalString = b.substring(0, b.size-4)
    }
    if(finalString == "") {
      finalString = b
    }
    return finalString
  }

}

File: Assignment4/task2.py
from itertools import combinations
from pyspark import SparkContext,SparkConf
import time
import sys
import queue
from collections import defaultdict
import copy

'''
case = int(sys.argv[1])
support = sys.argv[2]
input_file_path = sys.argv[3]
output_file_path = sys.argv[4]
'''

threshold = 7
input_file_path = "/Users/gopi/Desktop/Assignment4/ub_sample_data.csv"
between_file_path = "/Users/gopi/Desktop/Assignment4/between.csv"
output_file_path = "/Users/gopi/Desktop/Assignment4/output2.csv"

conf = SparkConf().setMaster("local[*]").setAppName("assign-4-task-1").set("spark.executor.memory", "4g").set("spark.driver.memory", "4g")
sc = SparkContext(conf=conf)
sc.setLogLevel("ERROR")


def get_betweeness(adjacency_list):
    result = defaultdict(lambda : 0)
    vertices = list(adjacency_list.keys())
    counts = [0 for i in range(13)]
    
    for vertex in vertices:
        neighbours = queue.Queue()
        nodes_in_queue = set()
        neighbours.put(vertex)
        nodes_in_queue.add(vertex)
        
        child_parent_map = defaultdict(lambda : set())
        shortest_paths_count = defaultdict(lambda : 0.0)
        shortest_paths_count[vertex] = 1.0
        levels_map = defaultdict(lambda : 0)
        levels_map[vertex] = 0
        
        while(not neighbours.empty()):
            current_node = neighbours.get()
            #nodes_in_queue.remove(current_node)
            next_level_nodes = adjacency_list[current_node]
            
            for neighbour in next_level_nodes:
                if neighbour not in levels_map:
                    levels_map[neighbour] = levels_map[current_node] + 1
                    shortest_paths_count[neighbour] = 1.0
                    
                    if neighbour not in nodes_in_queue:
                        neighbours.put(neighbour)
                        nodes_in_queue.add(neighbour)
                    
                    child_parent_map[neighbour].add(current_node)
                else:
                    if levels_map[neighbour] == levels_map[current_node] + 1:
                        child_parent_map[neighbour].add(current_node)
                        shortest_paths_count[neighbour] += 1.0
        
        #print(shortest_paths_count)
        betweeness = get_edge_betweeness(child_parent_map, shortest_paths_count, levels_map, vertex)
        a = set()
        [a.add(i) for i in levels_map.values()]
        counts[len(a)] +=1
        
        for edge in list(betweeness.keys()):
            result[edge] = result[edge] + (betweeness[edge] / 2.0)
    
    #print(result[("0QREkWHGO8-Z_70qx1BIWw","kKTcYPz47sCDH1_ylnE4ZQ")])
    result = list(sorted(result.items(), key = lambda x : (-x[1], x[0])))
    print(counts)
    return result
    
    
def get_edge_betweeness(child_parent_map, shortest_paths_count, levels_map, root_node):
    node_score = defaultdict(lambda : 1) #node_credits
    betweeness = {}    
    vertices_reverse = sorted(levels_map.keys(), key = lambda x : levels_map[x], reverse=True)
    

    for node in vertices_reverse:
        if node != root_node:
            share_count = []
            for parent in child_parent_map[node]:
                share_count.append(shortest_paths_count[parent])
            share = sum(share_count)     
            for parent in child_parent_map[node]:
                key = tuple(sorted([parent, node]))
                betweeness[key] = (node_score[node] * shortest_paths_count[parent]) / share
                node_score[parent] += betweeness[key]
    
    return betweeness
        

def get_communities(adjacency_list):
    communities = list()
    vertex_set = set(adjacency_list.keys())
    
    while len(vertex_set) > 0:
        vertices_list = list(vertex_set)
        first_node = vertices_list[0]
        visited_nodes = set()
        
        neighbours = queue.Queue()
        nodes_in_queue = set()
        neighbours.put(first_node)
        nodes_in_queue.add(first_node)
        
        while(not neighbours.empty()):
            vertex = neighbours.get()
            visited_nodes.add(vertex)
            #nodes_in_queue.remove(vertex)
            
            adjacent_nodes = adjacency_list[vertex]
            for adjacent_node in adjacent_nodes:
                if adjacent_node not in visited_nodes and adjacent_node not in nodes_in_queue:
                    neighbours.put(adjacent_node)
                    nodes_in_queue.add(adjacent_node)
            
        community_list = sorted(list(visited_nodes))
        communities.append(community_list)
        
        vertex_set = vertex_set.difference(visited_nodes)
    
    return communities

def get_modularity(adjacency_list, community_list, m):
    
    result = 0
    for community in community_list:
        #comb = combinations(community, 2)
        for i1 in community:
            for i2 in community:
                key = tuple(sorted([i1,i2]))
                length_1 = len(adjacency_list[i1])
                length_2 = len(adjacency_list[i2])
            
                value = 1 if key in data_set else 0
                community_mod = value - ((length_1 * length_2) / (2.0 *m))
            #print(community_mod)
                result += community_mod
    return result / (2.0*m)
    
def write_betweeness_data_to_file(output_file_path, result):
    file = open(output_file_path, 'w')
    
    for data in result:
        key = data[0]
        value = data[1]
        string = "('" + key[0] + "'" + ", '" + key[1] + "')," + str(round(value, 5))
        file.write(string)
        file.write("\n")   
    file.close() 
    
def write_data_to_file(output_file_path, results):
    with open(output_file_path, 'w') as file:
        for data in results:
          final_str = str(data)[1:-1]
          file.write(final_str)
          file.write("\n")
    file.close()
    
initial_rdd = sc.textFile(input_file_path)
first = initial_rdd.first()
initial_rdd = initial_rdd.filter(lambda x : x != first).map(lambda line: line.split(","))

user_business_map = initial_rdd.map(lambda x : (x[0], x[1])).groupByKey().mapValues(lambda x : set(sorted(list(x)))).collectAsMap()
user_id_list = list(user_business_map.keys())
user_comb = list(combinations(user_id_list, 2))
data_set = set()

adjacent_nodes = defaultdict(lambda x : set())

for comb in user_comb:
    set_1 = user_business_map[comb[0]]
    set_2 = user_business_map[comb[1]]
    
    if len(set_1.intersection(set_2)) >= threshold:
        if comb[0] not in adjacent_nodes:
            adjacent_nodes[comb[0]] = set([comb[1]])
        else:
            adjacent_nodes[comb[0]].add(comb[1])
        if comb[1] not in adjacent_nodes:
            adjacent_nodes[comb[1]] = set([comb[0]])
        else:
            adjacent_nodes[comb[1]].add(comb[0])
        data_set.add(tuple(sorted(comb)))
        

#print(adjacent_nodes)

betweeness = get_betweeness(adjacent_nodes)#edg
write_betweeness_data_to_file(between_file_path, betweeness)

m = len(betweeness)
#data_set = initial_rdd.map(lambda x : tuple(sorted((x[0], x[1])))).collect()
total_edges = len(betweeness)

communities = get_communities(adjacent_nodes)
modularity = get_modularity(adjacent_nodes, communities, m)
adjacent_nodes_copy = copy.deepcopy(adjacent_nodes)
print("modu is:",modularity)
while total_edges > 0:
    h_edge_val = betweeness[0]
    h_val = h_edge_val[1]
    h_edge = h_edge_val[0]
    
    for edge in betweeness:
        temp = edge[0]
        if edge[1] == h_val:
            adjacent_nodes_copy[temp[0]].remove(temp[1])
            adjacent_nodes_copy[temp[1]].remove(temp[0])
            total_edges -= 1
    #print(total_edges)
            
    communities_modified = get_communities(adjacent_nodes_copy)
    modularity_modified = get_modularity(adjacent_nodes, communities_modified, m)
    modularity = modularity_modified if modularity_modified >= modularity else modularity
    print(modularity)
    communities = communities_modified if modularity_modified >= modularity else communities
    betweeness = get_betweeness(adjacent_nodes_copy)
    
results = sorted(communities, key = lambda x : (len(x), x))
write_data_to_file(output_file_path, results)
#print(results)
